File: src/analysis.test.ts
import { expect, test } from "bun:test";
import { hrAtPace, splitShape, splitTable, strokeSummary } from "./analysis.ts";
import { projectGoal } from "./commands/stats.ts";
import type { StrokeData, Workout, WorkoutSplit } from "./models.ts";
import type { GoalProgress } from "./stats.ts";

function split(timeTenths: number, distance: number, spm: number, hr: number): WorkoutSplit {
  return {
    type: "distance",
    time: timeTenths,
    distance,
    stroke_rate: spm,
    heart_rate: { average: hr, max: hr + 8 },
  };
}

function workoutWithSplits(splits: WorkoutSplit[]): Workout {
  return {
    id: 1,
    user_id: 1,
    date: "2026-07-03 12:00:00",
    distance: splits.reduce((s, x) => s + (x.distance ?? 0), 0),
    type: "rower",
    time: splits.reduce((s, x) => s + x.time, 0),
    time_formatted: "34:39.1",
    workout: { targets: { pace: 1750 }, splits },
  };
}

const NEGATIVE_SPLIT_WORKOUT = workoutWithSplits([
  split(4280, 1200, 24, 93),
  split(4210, 1200, 25, 107),
  split(4140, 1200, 25, 115),
  split(4076, 1200, 26, 120),
  split(4088, 1200, 26, 122),
]);

test("splitTable computes per-split pace from tenths", () => {
  const rows = splitTable(NEGATIVE_SPLIT_WORKOUT);
  expect(rows.length).toBe(5);
  expect(rows[0]!.pace_500m_seconds).toBe(178.3);
  expect(rows[0]!.pace_500m).toBe("2:58.3");
  expect(rows[3]!.pace_500m).toBe("2:49.8");
  expect(rows[0]!.hr_avg).toBe(93);
  expect(rows[0]!.hr_max).toBe(101);
});

test("splitShape detects the 07/03-style negative split", () => {
  expect(splitShape(splitTable(NEGATIVE_SPLIT_WORKOUT))).toBe("negative");
});

test("splitShape detects even, positive, variable, and unknown", () => {
  const even = workoutWithSplits([
    split(4200, 1200, 25, 110),
    split(4210, 1200, 25, 112),
    split(4195, 1200, 25, 113),
    split(4205, 1200, 25, 114),
  ]);
  expect(splitShape(splitTable(even))).toBe("even");

  const positive = workoutWithSplits([
    split(4000, 1200, 26, 118),
    split(4100, 1200, 25, 120),
    split(4250, 1200, 24, 121),
    split(4350, 1200, 23, 122),
  ]);
  expect(splitShape(splitTable(positive))).toBe("positive");

  const variable = workoutWithSplits([
    split(4000, 1200, 26, 118),
    split(4400, 1200, 22, 110),
    split(3950, 1200, 27, 125),
    split(4380, 1200, 22, 112),
  ]);
  expect(splitShape(splitTable(variable))).toBe("variable");

  expect(splitShape(splitTable(workoutWithSplits([split(4200, 1200, 25, 110)])))).toBe("unknown");
  expect(splitShape([])).toBe("unknown");
});

test("strokeSummary aggregates samples", () => {
  const strokes: StrokeData[] = [
    { t: 100, d: 500, p: 1750, spm: 24, hr: 110 },
    { t: 200, d: 1000, p: 1730, spm: 25, hr: 118 },
    { t: 300, d: 1500, p: 1710, spm: 26, hr: 126 },
    { t: 400, d: 2000, p: 0, spm: 0, hr: 0 },
  ];
  const s = strokeSummary(strokes);
  expect(s.samples).toBe(4);
  expect(s.avg_pace_500m_seconds).toBe(173);
  expect(s.avg_pace_500m).toBe("2:53.0");
  expect(s.avg_spm).toBe(25);
  expect(s.avg_hr).toBe(118);
  expect(s.max_hr).toBe(126);
});

function goalFixture(overrides: Partial<GoalProgress>): GoalProgress {
  return {
    target: 1_000_000,
    totalMeters: 900_000,
    progress: 0.9,
    weeksElapsed: 26,
    totalWeeks: 52,
    remainingMeters: 100_000,
    remainingWeeks: 26,
    requiredPace: 3846,
    currentAvgPace: 20_000,
    onPace: true,
    ...overrides,
  };
}

const PROJECT_NOW = new Date("2026-07-06T12:00:00");

function daysAfterNow(n: number): Date {
  return new Date(PROJECT_NOW.getTime() + n * 24 * 60 * 60 * 1000);
}

test("projectGoal projects over actual remaining time", () => {
  const active = projectGoal(goalFixture({}), daysAfterNow(26 * 7), PROJECT_NOW);
  expect(active.projected_total_meters).toBe(900_000 + 26 * 20_000);
  expect(active.shortfall_meters).toBe(0);
});

test("projectGoal adds nothing after the goal window ends", () => {
  const expired = projectGoal(
    goalFixture({ weeksElapsed: 60, remainingWeeks: 1 }),
    daysAfterNow(-2),
    PROJECT_NOW,
  );
  expect(expired.projected_total_meters).toBe(900_000);
  expect(expired.projected_pct).toBe(90);
  expect(expired.shortfall_meters).toBe(100_000);
});

test("projectGoal handles fractional final weeks without over-projecting", () => {
  const halfWeek = projectGoal(goalFixture({}), daysAfterNow(3.5), PROJECT_NOW);
  expect(halfWeek.projected_total_meters).toBe(900_000 + 10_000);
});

function steadyWorkout(id: number, date: string, paceSecs: number, hr: number): Workout {
  const distance = 6000;
  return {
    id,
    user_id: 1,
    date,
    distance,
    type: "rower",
    time: Math.round(paceSecs * (distance / 500) * 10),
    time_formatted: "x",
    heart_rate: { average: hr },
  };
}

const NOW = new Date("2026-07-05T12:00:00");

test("hrAtPace buckets steady work and reports early-late drift", () => {
  const workouts: Workout[] = [
    steadyWorkout(1, "2026-05-20 08:00:00", 173, 120),
    steadyWorkout(2, "2026-05-27 08:00:00", 174, 118),
    steadyWorkout(3, "2026-06-24 08:00:00", 172, 112),
    steadyWorkout(4, "2026-07-01 08:00:00", 171, 110),
    steadyWorkout(5, "2026-07-03 08:00:00", 168, 115),
    { ...steadyWorkout(6, "2026-07-02 08:00:00", 150, 140), rest_time: 600 },
    steadyWorkout(7, "2026-01-01 08:00:00", 173, 130),
  ];
  const bands = hrAtPace(workouts, NOW, 8);

  expect(bands.length).toBe(2);
  const band170 = bands.find((b) => b.band_start_seconds === 170)!;
  expect(band170.workouts).toBe(4);
  expect(band170.early_avg_hr).toBe(119);
  expect(band170.late_avg_hr).toBe(111);
  expect(band170.hr_delta).toBe(-8);
  expect(band170.band).toBe("2:50.0–2:55.0");

  const band165 = bands.find((b) => b.band_start_seconds === 165)!;
  expect(band165.workouts).toBe(1);
  expect(band165.hr_delta).toBeNull();
});

File: internal/analysis/analysis_test.go
package analysis

import (
	"math"
	"testing"
	"time"

	"github.com/richhaase/c2/internal/models"
)

func intPtr(v int) *int { return &v }

func floatPtr(v float64) *float64 { return &v }

func split(timeTenths float64, distance, spm, hr int) models.WorkoutSplit {
	return models.WorkoutSplit{
		Type:       "distance",
		Time:       timeTenths,
		Distance:   intPtr(distance),
		StrokeRate: intPtr(spm),
		HeartRate:  &models.HeartRate{Average: intPtr(hr), Max: intPtr(hr + 8)},
	}
}

func workoutWithSplits(splits []models.WorkoutSplit) models.Workout {
	distance := 0
	totalTime := 0.0
	for _, s := range splits {
		if s.Distance != nil {
			distance += *s.Distance
		}
		totalTime += s.Time
	}
	return models.Workout{
		ID:            1,
		UserID:        1,
		Date:          "2026-07-03 12:00:00",
		Distance:      distance,
		Type:          "rower",
		Time:          int(totalTime),
		TimeFormatted: "34:39.1",
		Workout: &models.WorkoutDetail{
			Targets: &models.WorkoutTargets{Pace: floatPtr(1750)},
			Splits:  splits,
		},
	}
}

func negativeSplitWorkout() models.Workout {
	return workoutWithSplits([]models.WorkoutSplit{
		split(4280, 1200, 24, 93),
		split(4210, 1200, 25, 107),
		split(4140, 1200, 25, 115),
		split(4076, 1200, 26, 120),
		split(4088, 1200, 26, 122),
	})
}

func TestSplitTableComputesPerSplitPaceFromTenths(t *testing.T) {
	rows := SplitTable(negativeSplitWorkout())
	if len(rows) != 5 {
		t.Fatalf("rows length = %d, want 5", len(rows))
	}
	if got := *rows[0].Pace500mSeconds; got != 178.3 {
		t.Errorf("rows[0].pace_500m_seconds = %v, want 178.3", got)
	}
	if got := *rows[0].Pace500m; got != "2:58.3" {
		t.Errorf("rows[0].pace_500m = %q, want %q", got, "2:58.3")
	}
	if got := *rows[3].Pace500m; got != "2:49.8" {
		t.Errorf("rows[3].pace_500m = %q, want %q", got, "2:49.8")
	}
	if got := *rows[0].HRAvg; got != 93 {
		t.Errorf("rows[0].hr_avg = %d, want 93", got)
	}
	if got := *rows[0].HRMax; got != 101 {
		t.Errorf("rows[0].hr_max = %d, want 101", got)
	}
}

func TestSplitShapeDetectsNegativeSplit(t *testing.T) {
	if got := SplitShape(SplitTable(negativeSplitWorkout())); got != ShapeNegative {
		t.Errorf("SplitShape = %q, want %q", got, ShapeNegative)
	}
}

func TestSplitShapeDetectsEvenPositiveVariableAndUnknown(t *testing.T) {
	even := workoutWithSplits([]models.WorkoutSplit{
		split(4200, 1200, 25, 110),
		split(4210, 1200, 25, 112),
		split(4195, 1200, 25, 113),
		split(4205, 1200, 25, 114),
	})
	if got := SplitShape(SplitTable(even)); got != ShapeEven {
		t.Errorf("even SplitShape = %q, want %q", got, ShapeEven)
	}

	positive := workoutWithSplits([]models.WorkoutSplit{
		split(4000, 1200, 26, 118),
		split(4100, 1200, 25, 120),
		split(4250, 1200, 24, 121),
		split(4350, 1200, 23, 122),
	})
	if got := SplitShape(SplitTable(positive)); got != ShapePositive {
		t.Errorf("positive SplitShape = %q, want %q", got, ShapePositive)
	}

	variable := workoutWithSplits([]models.WorkoutSplit{
		split(4000, 1200, 26, 118),
		split(4400, 1200, 22, 110),
		split(3950, 1200, 27, 125),
		split(4380, 1200, 22, 112),
	})
	if got := SplitShape(SplitTable(variable)); got != ShapeVariable {
		t.Errorf("variable SplitShape = %q, want %q", got, ShapeVariable)
	}

	single := workoutWithSplits([]models.WorkoutSplit{split(4200, 1200, 25, 110)})
	if got := SplitShape(SplitTable(single)); got != ShapeUnknown {
		t.Errorf("single-split SplitShape = %q, want %q", got, ShapeUnknown)
	}
	if got := SplitShape(nil); got != ShapeUnknown {
		t.Errorf("empty SplitShape = %q, want %q", got, ShapeUnknown)
	}
}

func TestStrokeSummaryAggregatesSamples(t *testing.T) {
	strokes := []models.StrokeData{
		{T: floatPtr(100), D: floatPtr(500), P: floatPtr(1750), SPM: floatPtr(24), HR: floatPtr(110)},
		{T: floatPtr(200), D: floatPtr(1000), P: floatPtr(1730), SPM: floatPtr(25), HR: floatPtr(118)},
		{T: floatPtr(300), D: floatPtr(1500), P: floatPtr(1710), SPM: floatPtr(26), HR: floatPtr(126)},
		{T: floatPtr(400), D: floatPtr(2000), P: floatPtr(0), SPM: floatPtr(0), HR: floatPtr(0)},
	}
	s := StrokeSummary(strokes)
	if s.Samples != 4 {
		t.Errorf("samples = %d, want 4", s.Samples)
	}
	if got := *s.AvgPace500mSeconds; got != 173 {
		t.Errorf("avg_pace_500m_seconds = %v, want 173", got)
	}
	if got := *s.AvgPace500m; got != "2:53.0" {
		t.Errorf("avg_pace_500m = %q, want %q", got, "2:53.0")
	}
	if got := *s.AvgSPM; got != 25 {
		t.Errorf("avg_spm = %v, want 25", got)
	}
	if got := *s.AvgHR; got != 118 {
		t.Errorf("avg_hr = %d, want 118", got)
	}
	if got := *s.MaxHR; got != 126 {
		t.Errorf("max_hr = %v, want 126", got)
	}
}

func steadyWorkout(id int64, date string, paceSecs float64, hr int) models.Workout {
	distance := 6000
	return models.Workout{
		ID:            id,
		UserID:        1,
		Date:          date,
		Distance:      distance,
		Type:          "rower",
		Time:          int(math.Round(paceSecs * (float64(distance) / 500) * 10)),
		TimeFormatted: "x",
		HeartRate:     &models.HeartRate{Average: intPtr(hr)},
	}
}

func TestHRAtPaceBucketsSteadyWorkAndReportsDrift(t *testing.T) {
	now := time.Date(2026, time.July, 5, 12, 0, 0, 0, time.Local)

	interval := steadyWorkout(6, "2026-07-02 08:00:00", 150, 140)
	interval.RestTime = intPtr(600)

	workouts := []models.Workout{
		steadyWorkout(1, "2026-05-20 08:00:00", 173, 120),
		steadyWorkout(2, "2026-05-27 08:00:00", 174, 118),
		steadyWorkout(3, "2026-06-24 08:00:00", 172, 112),
		steadyWorkout(4, "2026-07-01 08:00:00", 171, 110),
		steadyWorkout(5, "2026-07-03 08:00:00", 168, 115),
		interval,
		steadyWorkout(7, "2026-01-01 08:00:00", 173, 130),
	}
	bands := HRAtPace(workouts, now, 8)

	if len(bands) != 2 {
		t.Fatalf("bands length = %d, want 2", len(bands))
	}

	band170 := findBand(t, bands, 170)
	if band170.Workouts != 4 {
		t.Errorf("band170.workouts = %d, want 4", band170.Workouts)
	}
	if got := *band170.EarlyAvgHR; got != 119 {
		t.Errorf("band170.early_avg_hr = %d, want 119", got)
	}
	if got := *band170.LateAvgHR; got != 111 {
		t.Errorf("band170.late_avg_hr = %d, want 111", got)
	}
	if got := *band170.HRDelta; got != -8 {
		t.Errorf("band170.hr_delta = %d, want -8", got)
	}
	if band170.Band != "2:50.0–2:55.0" {
		t.Errorf("band170.band = %q, want %q", band170.Band, "2:50.0–2:55.0")
	}

	band165 := findBand(t, bands, 165)
	if band165.Workouts != 1 {
		t.Errorf("band165.workouts = %d, want 1", band165.Workouts)
	}
	if band165.HRDelta != nil {
		t.Errorf("band165.hr_delta = %v, want nil", *band165.HRDelta)
	}
}

func findBand(t *testing.T, bands []HRPaceBand, start int) HRPaceBand {
	t.Helper()
	for _, b := range bands {
		if b.BandStartSeconds == start {
			return b
		}
	}
	t.Fatalf("no band with band_start_seconds = %d", start)
	return HRPaceBand{}
}
